[WIP] #4 - webhookAPI

fanwashere opened the pull request.

fanwashere committed:
[WIP] #4 - webhookAPI

fanwashere committed:
[WIP] #4 - webhookAPI

Completed API, still needs to write more detailed tests.

fanwashere committed:
[WIP] #4 - webhookAPI

For some reason, the AbstractJson.scala file was not added. There is an added seq method for converting Json arrays to Seq objects. Also removed copy+paste functions from WebhookOp.scala.

fanwashere committed:
[WIP] #4 - webhookAPI

Added class WebhookUpdateInput. This had to be a seperate class from WebhookInput because createWebhook has two required parameters while updateWebhook contain only optional parameters.

In addition, changed tests for updateWebhook to accommodate the fact that add_events and remove_events cannot be used in the same method call. Comment was added to the code to document this fact.

fanwashere commented:
@shunjikonishi review me.

shunjikonishi commented on src/main/scala/codecheck/github/models/Webhook.scala:
```diff
@@ -0,0 +1,48 @@
+}
+
+case class WebhookConfig(
+  val url: String,
```
val is not necessary

shunjikonishi commented on src/main/scala/codecheck/github/models/Webhook.scala:
```diff
@@ -0,0 +1,48 @@
+
+case class WebhookUpdateInput(
+  config: Option[WebhookConfig] = None,
+  events: Option[Seq[String]] = None,
```
Option is redundncy

fanwashere replied:
Do you mean instead of Option[Seq[String]] = None, use Seq[String] = Nil?

shunjikonishi commented on src/main/scala/codecheck/github/models/Webhook.scala:
```diff
@@ -0,0 +1,48 @@
+  def message = get("message")
+}
+
+case class WebhookInput(
```
If you want to separate input class, I think WebhookCreateInput is better.

shunjikonishi commented on src/main/scala/codecheck/github/operations/WebhookOp.scala:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  def getWebhook(owner: String, repo: String, id: Long): Future[Option[Webhook]] = {
+  	self.exec("GET", s"/repos/${owner}/${repo}/hooks/${id}").map { res =>
```
If you want to handle 404 in map function, fail404=false is necessary.

shunjikonishi commented on src/main/scala/codecheck/github/operations/WebhookOp.scala:
```diff
@@ -0,0 +1,81 @@
+  	}
+  } 
+
+  def createWebhook(owner: String, repo: String, input: WebhookInput): Future[Option[Webhook]] = {
```
Why Option?
I think this API never return 404.

fanwashere replied:
It will return 404 if organization or repo is invalid.

sukeshni replied:
@fanwashere 
I think it should return NotFoundException, are sure about 404 ?

fanwashere replied:
This was changed again later to fail404=true. I get some sort of error when I put in invalid info (this is the intended effect). It may be NotFoundException.

shunjikonishi commented on src/main/scala/codecheck/github/operations/WebhookOp.scala:
```diff
@@ -0,0 +1,81 @@
+
+  //It is apparently an issue with Github's Webhook API that add_events and remove_events cannot be done
+  //in a single operation. To add and remove events, it must be done through two seperate calls of updateWebhook.
+  def updateWebhook(owner: String, repo: String, id: Long, input: WebhookUpdateInput): Future[Option[Webhook]] = {
```
Also why option?

fanwashere replied:
If using invalid organization or repo, it will 404. Normally it returns the updated Webhook, but when 404 it should return None.

shunjikonishi commented on src/main/scala/codecheck/github/models/AbstractJson.scala:
```diff
@@ -59,4 +60,16 @@ class AbstractJson(value: JValue) {
+    }
+  }
+
+  def seq(path: String): Seq[String] = seqOpt(path)
```
About seq and seqOpt, I think there might be better implementation.
We'll consider later.

You don't have to fix it in this branch.

fanwashere replied:
I agree. I am not very experienced with Scala so I just went with the first thing I thought of.

shunjikonishi commented on src/test/scala/WebhookOpSpec.scala:
```diff
@@ -0,0 +1,112 @@
+	    	assert(result.length > 0)
+	    }
+
+      it("should fail if there are no webhookswith valid owner, repo.") {
```
Maybe this test is not completed.
Just copied above.

shunjikonishi commented on src/test/scala/WebhookOpSpec.scala:
```diff
@@ -0,0 +1,112 @@
+
+class WebhookOpSpec extends FunSpec with Constants with BeforeAndAfter {
+
+	val gURL = generateRandomWord;
```
It is better to use valid github webhook application.
You can use 

http://github-hook.herokuapp.com/hook

shunjikonishi commented:
I've commented.

@fanwashere fix me

fanwashere committed:
[WIP] #4 - webhookAPI

Made several changes after Koni-san's commments.

fanwashere committed:
[WIP] #4 - webhookAPI

Changed createWebhook and updateWebhook to fail on 404 rather than return None.

fanwashere commented:
Fixed.

@shunjikonishi review me.

shunjikonishi commented:
@sukeshni review first.

I'll review after you.

sukeshni commented:
@fanwashere Indentation, other then that LGTM.

fanwashere committed:
[WIP] #4 - webhookAPI

Changed all indentation to spaces. No change in code.

fanwashere commented:
I think @shunjikonishi wants to review it before merge. Please review.

shunjikonishi commented:
@fanwashere fix me
Sorry for my lack of explain.

Your change 
(WebHookInput, WebHookUpdateInput) -> (WebHookCreateInput, WebHookUpdateInput)

I meant 
(WebHookCreateInput, WebHookUpdateInput)

Otherwise LGTM.

fanwashere committed:
[WIP] #4 - webhookAPI

Changed the name of the class WebhookInput -> WebhookCreateInput for clarity. Now there are two Input classes, one for createWebhook (WebhookCreateInput) and one for updateWebhook (WebhookUpdateInput). All instances of usage of the class name was replaced with the new name.

No other changes were made to the code.

fanwashere commented:
@shunjikonishi review me.

I made the change in the class name as you suggested. I hope I got it right this time!

fanwashere committed:
[WIP] #4 - webhookAPI

Fixed indentation level for consistency.

shunjikonishi commented:
![LGTM](http://lgtm.in/p/8wkMhYhXp)

fanwashere committed:
Merge branch 'master' into 4-webhookAPI

Conflicts:
	src/main/scala/codecheck/github/api/GitHubAPI.scala

  Resolved.
